src/lib/utils/sqlite_wrapper.cpp

```diff
@@ -19,27 +19,222 @@
+      }
+
+      case SQLITE_NULL: {
+        row.emplace_back(NULL_VALUE);
+        break;
+      }
+
+      case SQLITE_BLOB:
+      default: {
+        sqlite3_finalize(sqlite_statement);
+        Fail("Column type not supported.");
+      }
+    }
+  }
+
+  table->append(row);
+}
 
 }  // namespace
 
 namespace opossum {
 
-SQLiteWrapper::SQLiteWrapper() {
+// We include the address of this wrapper in the URI for the sqlite db so that two wrappers don't interfere
+// See also https://www.sqlite.org/inmemorydb.html, starting at "If two or more distinct but shareable
+// in-memory databases"
+SQLiteWrapper::SQLiteWrapper()
+    : _uri{std::string{"file:sqlitewrapper_"} + std::to_string(reinterpret_cast<uintptr_t>(this)) +  // NOLINT
+           "?mode=memory&cache=shared"},
+      main_connection{_uri} {
+  Assert(sqlite3_threadsafe(), "Expected sqlite to be compiled with thread-safety");
+}
+
+SQLiteWrapper::Connection::Connection(const std::string& uri) {
   // Explicity set parallel mode. On Linux it seems to be the default, on Mac, it seems to make a difference.
   const auto ret =
-      sqlite3_open_v2(":memory:", &_db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX, nullptr);  // NOLINT
+      sqlite3_open_v2(uri.c_str(), &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX,  // NOLINT
+                      nullptr);
   if (ret != SQLITE_OK) {
-    sqlite3_close(_db);
-    Fail("Cannot open database: " + std::string(sqlite3_errmsg(_db)));
+    sqlite3_close(db);
+    Fail("Cannot open database: " + std::string(sqlite3_errmsg(db)));
   }
 
   // Make LIKE case sensitive, just like in Hyrise
-  raw_execute_query("PRAGMA case_sensitive_like = true");
+  raw_execute_query("PRAGMA case_sensitive_like = true;");
+}
+
+SQLiteWrapper::Connection::Connection(Connection&& other) noexcept : db(other.db) { other.db = nullptr; }
+
+SQLiteWrapper::Connection::~Connection() {
+  if (db) {
+    sqlite3_close(db);
+  }
+}
+
+SQLiteWrapper::Connection SQLiteWrapper::new_connection() const { return Connection{_uri}; }
+
+std::shared_ptr<Table> SQLiteWrapper::Connection::execute_query(const std::string& sql) const {
+  sqlite3_stmt* sqlite_statement;
+
+  auto sql_pipeline = SQLPipelineBuilder{sql}.create_pipeline();
+  const auto& queries = sql_pipeline.get_sql_per_statement();
+
+  // We need to split the queries such that we only create columns/add rows from the final SELECT query
+  std::vector<std::string> queries_before_select(queries.begin(), queries.end() - 1);
+  std::string select_query = queries.back();
+
+  int rc;
+  for (const auto& query : queries_before_select) {
+    rc = sqlite3_prepare_v2(db, query.c_str(), -1, &sqlite_statement, nullptr);
+
+    if (rc != SQLITE_OK) {
+      sqlite3_finalize(sqlite_statement);
+      Fail("Failed to execute query \"" + query + "\": " + std::string(sqlite3_errmsg(db)) + "\n");
+    }
+
+    while ((rc = sqlite3_step(sqlite_statement)) != SQLITE_DONE) {
```

Comment:
Do you know which steps are performed here until `SQLITE_DONE`?

Reply:
What steps? This is SQLITE_ROW until we reach the end of the table.

Reply:
I was just curious. The SQLite doc isn't too bad. Should have looked that up first.